File: backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ApiKeysService } from './api-keys.service';

@Injectable()
export class AuthService {
  constructor(
    private readonly jwtService: JwtService,
    private readonly apiKeysService: ApiKeysService
  ) {}

  issueToken(clientId: string, clientSecret: string) {
    const adminClientId = process.env.AUTH_ADMIN_CLIENT_ID || '';
    const adminClientSecret = process.env.AUTH_ADMIN_CLIENT_SECRET || '';

    if (clientId !== adminClientId || clientSecret !== adminClientSecret) {
      throw new UnauthorizedException('Invalid client credentials');
    }

    const payload = { sub: clientId, role: 'admin' } as const;
    return this.issueAdminToken(payload.sub, payload.role);
  }

  issueAdminToken(username: string, role: 'admin' | 'editor') {
    return {
      accessToken: this.jwtService.sign({ sub: username, role }),
      expiresIn: Number(process.env.AUTH_JWT_TTL || 3600)
    };
  }

  issueTenantToken(tenantId: string, username: string) {
    return {
      accessToken: this.jwtService.sign({ sub: username, role: 'tenant', tenantId }),
      expiresIn: Number(process.env.AUTH_JWT_TTL || 3600)
    };
  }

  async validateJwt(token: string) {
    return this.jwtService.verifyAsync(token);
  }

  async validateApiKey(apiKey: string) {
    return this.apiKeysService.validate(apiKey);
  }
}

File: backend-java/src/main/java/com/neria/manager/auth/AuthService.java
package com.neria.manager.auth;

import com.neria.manager.common.entities.ApiKey;
import com.neria.manager.common.security.AuthContext;
import com.neria.manager.common.services.JwtService;
import com.neria.manager.config.AppProperties;
import io.jsonwebtoken.Claims;
import java.util.HashMap;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class AuthService {
  private final JwtService jwtService;
  private final ApiKeysService apiKeysService;
  private final AppProperties properties;

  public AuthService(JwtService jwtService, ApiKeysService apiKeysService, AppProperties properties) {
    this.jwtService = jwtService;
    this.apiKeysService = apiKeysService;
    this.properties = properties;
  }

  public TokenResult issueToken(String clientId, String clientSecret) {
    String adminClientId = properties.getAuth().getAdminClientId();
    String adminClientSecret = properties.getAuth().getAdminClientSecret();
    if (!adminClientId.equals(clientId) || !adminClientSecret.equals(clientSecret)) {
      throw new IllegalArgumentException("Invalid client credentials");
    }
    return issueAdminToken(clientId, "admin");
  }

  public TokenResult issueAdminToken(String username, String role) {
    Map<String, Object> claims = new HashMap<>();
    claims.put("sub", username);
    claims.put("role", role);
    String token = jwtService.sign(claims);
    return new TokenResult(token, jwtService.getTtlSeconds());
  }

  public TokenResult issueTenantToken(String tenantId, String username) {
    Map<String, Object> claims = new HashMap<>();
    claims.put("sub", username);
    claims.put("role", "tenant");
    claims.put("tenantId", tenantId);
    String token = jwtService.sign(claims);
    return new TokenResult(token, jwtService.getTtlSeconds());
  }

  public AuthContext validateJwt(String token) {
    Claims claims = jwtService.verify(token);
    return AuthContext.builder()
        .type("jwt")
        .sub(String.valueOf(claims.get("sub")))
        .role(String.valueOf(claims.get("role")))
        .tenantId(claims.get("tenantId") != null ? String.valueOf(claims.get("tenantId")) : null)
        .build();
  }

  public AuthContext validateApiKey(String apiKey) {
    ApiKey record = apiKeysService.validate(apiKey);
    if (record == null) {
      throw new IllegalArgumentException("Invalid API key");
    }
    return AuthContext.builder()
        .type("apiKey")
        .tenantId(record.getTenantId())
        .apiKeyId(record.getId())
        .build();
  }

  public record TokenResult(String accessToken, long expiresIn) {}
}
